### ISSUE TITLE
Tables get registered twice sometimes

### ISSUE THREAD

**Issue by juanignaciosl** [CartoDB/cartodb#2068]
Some table metadata is duplicated. You can find what tables with this query:
```
select u.id, u.username, t.min, t.max, t.diff, t.name, t.table_id, t.c, t.di from users u, (select count(1) c, min(updated_at) min, max(updated_at) max, max(updated_at) - min(updated_at) as diff, name, array_agg(data_import_id) di, user_id, table_id from user_tables where updated_at > '2014-8-01' group by name, user_id, table_id having count(1) > 1 limit 1000) t where u.id = t.user_id order by max desc;
```

It looks like it's because of ghost tables: duplication begun after ghost tables activation, and duplicated tables have (also) been registered with ghost table mechanism.

We've found a [probable race condition when ghost tables are created](https://github.com/CartoDB/cartodb/blob/07ad0ea4d12768d51621ba1a174cefeb9392d402/app/models/user.rb#L1109): CartoDB saves the new "real table" without checking whether table already exists, so if a table is returned by `search_for_cartodbfied_tables` method (that only returns _unregistered_ tables indeed) and it's registered by, for example, the importer before [ghost tables registration](https://github.com/CartoDB/cartodb/blob/07ad0ea4d12768d51621ba1a174cefeb9392d402/app/models/user.rb#L1109) it will be registered twice.

The same can happen in the opposite direction: importer might register a table that has already been registered by ghost tables. In that case importer should update table metadata with data_import_id.

So, needed changes:
- ~~Ghost tables: check for existence just before saving a new instance. Don't save if it exists.~~
- ~~Normal table registration: check for existence before saving a new instance. If it already exists, update with new values.~~
- [X] Avoid triggering ghost table registration twice.

Cleanup of duplicated data is postponed at #2213 until checking this fixes the problem.

cc @javisantana @Kartones

**Comment by javisantana** [comment]
> Create a db constraint that avoids duplication.

related to this I'd think what to do with the current already duplicated tables.

**Comment by Kartones** [comment]
The only issue I see here is: 
> Ghost tables: check for existence just before saving a new instance. Don't save if it exists.
How do we know is the same table the one that just appeared? Will we just trust is the same ghost tables is going to register? or should we do some simple column names check to see if it at least looks the same?

Apart from that +1, it is not the first race condition of importer + ghost tables working side by side, but if can be solved easily, perfect.

**Comment by juanignaciosl** [comment]
@santana: I'll think about it deeper later, but...
* for those that have data_import_id in a duplication it seems clear that row is better and the other one, created by ghost table registration, should've never existed.
* for those that haven't data_import_id I'd merge rows, probably just overriding older with newer and removing the older row. Some of the tables are from team members, I'll check with them first.

@Kartones: don't you think we can just trust table name? Since it's a PostgreSQL physical name the only way it can be different is if table was imported, renamed and user imported/created a new one with the same before first one was found, which seems quite unlikely. I think a column check isn't too useful, since two tables with the same name, created very closely in time, are likely to have the same structure.

**Comment by Kartones** [comment]
Yes, is an unlikely scenario, I was jus thinking (perhaps too quickly) about something like trying to import multiple times the same "test.csv" file "at once" while renaming an existing table to "test" also, or things like that, but is true it is more intentional screwing that normal usage, so for the time being just forget it.

**Comment by juanignaciosl** [comment]
@Kartones : just realised table `oid` is stored at `user_tables`, I think that's a value we can certainly trust, isn't it? I've updated the test query, it's the same for current duplicated tables (adding `oid` to grouping doesn't add new rows).

**Comment by javisantana** [comment]
there are other cases that you will find: same oid with different name and same name with different oid. Maybe this help you to create the migration

**Comment by juanignaciosl** [comment]
@javisantana I'm pretty sure we did a wrong problem statement (and obviously the proposed solution). I haven't been able to reproduce the problem because table.validate checks for uniqueness of `(user_id, table_name)`. It's a race condition with `validate` and `save`, and the only actual solution for that would be adding the constraint and catching the error. Nevertheless, after taking a look at the (incredibly useful) debug information of ghost tables at Rollbar, I've seen registration is enqueued multiple times for the same user. Example (faked name):

```
2015-02-11 04:47 am	que05.useb.cartodb.net	registering table dolphin_kisses
2015-02-11 04:47 am	que04.useb.cartodb.net	registering table dolphin_kisses
```

So we're enqueuing this multiple times. When I talked to you you said the intent was running this when a user enters the dashboard, but it's not what this is actually doing.

In `./app/controllers/api/application_controller.rb` there's this code:
```
def check_ghost_tables
  params[:controller] == "api/json/visualizations" and params[:action] == "index"
end

...

if check_ghost_tables && current_user.search_for_modified_table_names
  if !current_user.link_ghost_tables_working
    ::Resque.enqueue(::Resque::UserJobs::SyncTables::LinkGhostTables, current_user.id)
```

BUT that file is not the controller + action of the `/dashboard` request, but _the api_. For example, requesting the dashboard makes two requests to that endpoint, which is also used for many other things as tag searchs and so on, so this is *really* likely to trigger multiple parallel ghost linking requests, thus increasing the chance of race conditions.

SO I propose moving that intent from `./app/controllers/api/application_controller.rb` to `./app/controllers/admin/visualizations_controller.rb`, which is the controller of the page instead of using the controller of the API. Maybe it'd be useful to create the constraint and catch the error as well, but I'm pretty sure this change will almost vanish this issue.

@Kartones , what do you think?

**Comment by Kartones** [comment]
Agreed, I'd put it as a before_filter only on some actions we want. 

But I advise you, the main method will still get called twice, as grabbing many things means same action, different parameters. I would also add more restrictions like "only if listing vis or tables" (but not tags, for example) <- Blind example without looking at which two requests are done

**Comment by juanignaciosl** [comment]
Currently requesting `/dashboard` page only triggers that method once. Keep in mind that's the _page_ load method, not any API endpoint. Indeed `Admin::VisualizationsController#index` returns html, I don't think there's any additional call to that method.

**Comment by Kartones** [comment]
Sorry, right, didn't saw it was the vis controller from /admin/, not from
/api/

Fine then!

**Comment by juanignaciosl** [comment]
FYI: no new duplicated tables since this fix, so I'll do #2213 ASAP.
